PlayLoop/PlayerInteractionHandler.js

```diff
@@ -0,0 +1,201 @@
+        if (!this.blockers.includes(reasonString)) {
+            this.blockers.push(reasonString);
+        }
+    }
+
+    static unblockInteractionsFor(reasonString) {
+        let index = this.blockers.indexOf(reasonString);
+        if (index > -1) {
+            this.blockers.splice(index, 1);
+        }
+    }
+
+    static isTargeting(obj) {
+        return this.currentTarget === obj;
+    }
+
+    static canTrigger(obj) {
+        return (this.isTargeting(obj) && this.blockers.length === 0);
+    }
+
+    static willAllowJump() {
+        let target = this.currentTarget || this.lastTarget;
+
+        if (!target) {
+            return true;
+        }
+
+        if (target.type === ObjectTypes.NPC) {
+            return false;
+        } 
+        if (target.type === ObjectTypes.DOOR) {
+            return (target.activationTrigger !== DOOR_TRIGGERS.UP);
+        }
+    }
+
+    static updateControl(id) {
+        // should probably merge updateControl stuff into Controller and read from there instead?
+        let buttonDown = Controller[id];
+        this.justPressed[id] = buttonDown && !this.pressed[id];
+        this.pressed[id] = buttonDown;
+    }
+
+    static handleInput() {
+        // should probably merge updateControl stuff into Controller and read from there instead?
+        // controls need to be updated every frame
+        this.updateControl('jump');
+        this.updateControl('up');
+        this.updateControl('down');
+
+        if (!this.currentTarget) {
+            return;
+        }
+        if (this.blockers.length > 0) {
+            this.recentlyBlocked = true;
+            return;
+        }
+
+        if (this.recentlyBlocked) {
+            // wait an extra frame before interacting if recently blocked
+            this.recentlyBlocked = false;
+            return;
+        }
+
+        if (this.currentTarget.type === ObjectTypes.NPC) {
+            if (this.justPressed.jump) {
+                this.currentTarget.startDialogue();
+            }
+        }
+        if (this.currentTarget.type === ObjectTypes.DOOR) {
+            let triggerType = this.currentTarget.activationTrigger;
+            let cueEntrance = false;
+
+            if (triggerType === DOOR_TRIGGERS.AUTOMATIC) {
+                cueEntrance = true;
+            }
+            else if (triggerType === DOOR_TRIGGERS.UP) {
+                cueEntrance = this.justPressed.up;
+            }
+            else if (triggerType === DOOR_TRIGGERS.DOWN) {
+                cueEntrance = this.justPressed.down;
+            }
+            
+            if (cueEntrance) {
+                this.currentTarget.sendPlayerThroughExit();
+            }
+        }
+    }
+
+    static updatePlayerTarget(player) {
+        if (player.death) {
+            this.lastTarget = this.currentTarget;
+            this.currentTarget = null;
+            return;
+        }
+
+        let playerX = player.x;
+        let playerY = player.y;
```

Comment:
i wouldn't create a variable for them, it just uses more memory, just use player.x and y down below

Reply:
Sure, that’s fine. I was taught to cache one-time work before loops to optimize for speed, but honestly A) it’s just lookups here so any gain would be tiny and B) optimizing either way here seems unlikely to matter given such a small list. Deleting those lines _will_ make the code a bit more concise without losing clarity though, so that’s a win.